In [3]:
 #use gpu
pip install tensorflow-gpu==2.0.0-rc1

     |████████████████████████████████| 380.5MB 42kB/s 
     |████████████████████████████████| 4.3MB 41.4MB/s 
     |████████████████████████████████| 501kB 47.4MB/s 


In [0]:
#imports
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [0]:
#train and preprocess data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

## In this task, you'll implement several ANN models with different batch sizes. Specifically:

Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 8 as the mini batch size.

In [6]:
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons
#set activation to relu since that seemed to work the best previously
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
#set loss to crossentropy since it previously performed better than hinge
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 8 as mini batch size
model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=False)
#evaluate 
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.07831397637828195
Test accuracy: 0.9783


Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 128 as the mini batch size.

In [7]:
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate 
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16336726671606303
Test accuracy: 0.9513


Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using the full sample as the batch.

In [8]:
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#use the full sample this time
model.fit(X_train, Y_train, batch_size=X_train.shape[0], epochs=20, verbose=False)
#evaluate 
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 2.232954284286499
Test accuracy: 0.1897


Compare the result of each model with each other. Which batch size did perform better?

The model with the mini batch size of 8 had the highest accuracy score, but this model is probably overfit because there was a large difference in accuracy scores between the test and training sets for this model. The model with the batch size set to 128 had a slightly lower accuracy test for the test set, but there was not as large of a difference between the training/test sets. Batch size 128 was therefore the best because of accuracy and a lack of overfitting.


## In this task, you'll implement several ANN models with different learning rates for the stochastic gradient descent. In all of the models below, use 128 as your mini batch size.

Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 0.01 as the learning rate.

In [0]:
from tensorflow.keras import optimizers
#set different learning rates to be used in the models 
sgd_001 = optimizers.SGD(lr=0.01)
sgd_100 = optimizers.SGD(lr=100)
sgd_00000001 = optimizers.SGD(lr=0.0000001)

In [10]:
model = Sequential()
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#learning rate set to .01
model.compile(optimizer=sgd_001, loss='categorical_crossentropy',
              metrics=['accuracy'])
#fit model
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16958927047848701
Test accuracy: 0.951


Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 100 as the learning rate.

In [11]:
model = Sequential()
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#learning rate set to 100
model.compile(optimizer=sgd_100, loss='categorical_crossentropy',
              metrics=['accuracy'])
#fit model
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: nan
Test accuracy: 0.098


Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 0.0000001 as the learning rate.

In [12]:
model = Sequential()
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#learning rate set to 0.0000001
model.compile(optimizer=sgd_00000001, loss='categorical_crossentropy',
              metrics=['accuracy'])
#fit model
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 2.4065140701293943
Test accuracy: 0.0988


Compare the result of each model with each other. Which learning rate did perform better?

The model with the learning rate set to 0.01 clearly performed the best. This model had an accuracy of 95% whereas the models with learning rates set to 100 and 0.0000001 had incredibly low accuracy rates (0.098 and 0.0988). 

